module_utils/network_lsr/argument_validator.py

```diff
@@ -596,6 +596,28 @@ def _validate_impl(self, value, name):
             )
         return Util.mac_ntoa(addr)
 
+class ArgValidatorClonedMac(ArgValidatorStr):
+    def __init__(self, name, force_len=None, required=False, default_value=None):
+        ArgValidatorStr.__init__(self, name, required, default_value, None)
+        self.force_len = force_len
+
+    def _validate_impl(self, value, name):
+        v = ArgValidatorStr._validate_impl(self, value, name)
+
+        if value in ["preserve", "permanent", "random", "stable"]:
+            return v
+
+        try:
+            addr = Util.mac_aton(v, self.force_len)
+        except MyError:
+            raise ValidationError(
+                name, "value '%s' is not a valid MAC address nor a valid mode ['preserve', 'permanent', 'random', 'stable']" % (value)
+            )
+        if not addr:
+            raise ValidationError(
```

Comment:
Maybe we should validate this before calling `Util.mac_aton()`, so that for the case of empty mac,  we can raise an exception earlier.

And in the NM, there is a default value `ethernet.cloned-mac-address=preserve`, maybe we should also set the default value preserve here. @thom311 , what do you think ?

Reply:
> And in the NM, there is a default value ethernet.cloned-mac-address=preserve, maybe we should also set the default value preserve here. @thom311 , what do you think ?

The default value of the ArgValidator should be chosen so, that it makes sense and so that it's most convenient to the callers of the code. The current `None` value is chosen, so that the callers can distinguish whether the value was explicitly set in YAML or whether it's unset (`None`). I don't know whether any of the callers actually uses that information.

Reply:
According a [gnome blog post](https://blogs.gnome.org/thaller/tag/clone-mac-address/) the default value was different on older version so I think we should not define a default

Reply:
The default value for `cloned-mac-address` is "unset" in the profile, which then allows to fallback to a connection default in NetworkManager.conf (if specified), and otherwise fallback to "preserve" (or "permanent" on earlier versions).

In any case, the default value in the profile is neither "preserve" nor "permanent". It only ends up effectively being one of those settings, if it's still unspecified in NetworkManager.conf.

As far as the role is concerned, it can implement whatever it wants and -- if the user does not specify cloned-mac-address in the playbook -- it may set the per profile value to "preserve". In practice, it makes however sense to follow what NetworkManager would do, and leave the value unset, if it's unset in the playbook...

In short, having `None` as default makes a lot of sense.